### ISSUE TITLE
False accept: a comma in a subscript with an index-returning pattern flag is accepted, which native Zsh rejects

### ISSUE THREAD

**Issue by ss-o** [z-shell/zsh-lint#387]
A comma inside a subscript carrying an *index-returning* pattern flag (`(i)`, `(I)`) is accepted, although native Zsh rejects it as `invalid subscript`. Found while probing the #382 fix; present on `main` today and independent of it — base and fixed binaries accept every row below identically.

**338 rows** of the #382 main probe.

### Minimized script

```zsh
typeset -A Z=( a 1 )
local -a m=( a1 abx abc )
print ${m[(i)a,b]}
```

`zsh -f -n` accepts this; `zsh -f` reports `invalid subscript` at evaluation. `zsh-lint` accepts it silently.

### The rule is flag-dependent

This is the part worth getting right, and it is why a blanket "no comma beside a flag" check would be wrong. A comma makes the subscript a range, and a range needs endpoints that are indices. A flag returning an *index* (`i`, `I`) cannot supply one; a flag returning a *value* or matching in place (`r`, `R`, `k`, `n:expr:`) can:

| Flag | `print ${m[(FLAG)a,b]}` |
| --- | --- |
| `(i)` | `invalid subscript` |
| `(I)` | `invalid subscript` |
| `(in:2:)` | `invalid subscript` |
| `(r)` | valid |
| `(R)` | valid |
| `(k)` | valid |
| `(ir)` | valid |
| `(n:2:)` | valid |

Note `(ir)`: combining `i` with `r` makes it valid again, so the test is on the flag combination's return kind, not on the presence of the letter `i`.

### Both oracles

Binaries from `main` at `07275f2` and the #382 branch at `91d6617`; identical verdicts. zsh 5.9.2. Every row here passes `zsh -f -n` and fails only at evaluation, so **`zsh -n` alone cannot see this family at all** — the runtime oracle is what exposes it.

| Source | Native | zsh-lint |
| --- | --- | --- |
| `print ${m[(i)a,b]}` | run: `invalid subscript` | accepted |
| `print ${m[(I)a,b]}` | run: `invalid subscript` | accepted |
| `v=${m[(i)a,b]}` | run: `invalid subscript` | accepted |
| `m[(i)a,b]=v` | run: `invalid subscript` | accepted |
| `print ${m[(i)${Z[a]},b]}` | run: `invalid subscript` | accepted |
| `m[(i)${Z[a]},b]=v` | run: `invalid subscript` | accepted |

### Rows that must keep parsing

| Source | Native | Current |
| --- | --- | --- |
| `print ${m[(r)a,b]}` | valid | accepted |
| `print ${m[(k)a,b]}` | valid | accepted |
| `print ${m[(ir)a,b]}` | valid | accepted |
| `print ${m[1,2]}` | valid | accepted |
| `print ${m[1,(i)a]}` | valid | accepted |
| `print ${m[(i)a]}` | valid | accepted |

The last two matter: a comma is fine in the same subscript as `(i)` when the flag applies to a *range endpoint* rather than to the whole range, so position within the subscript is part of the rule.

Method and probe generators: `docs/project/2026-09-24-flag-pattern-cut-guard-survey.md`.

**Comment by ss-o** [comment]
## Overlaps #389, and #387's rule is the better one

#389 (filed later, by me) reports the same family from a different angle, and #387's framing is the more accurate of the two.

#387 says the test is on the **flag combination's return kind**, and gives `(ir)` as valid while `(i)` is not — so the letter `i` alone is not the predicate. #389 says the rule is positional (first range endpoint vs second). A fix written to #389's description alone would get `(ir)` wrong, so #387's framing is the more useful starting point.

But neither issue's rule is complete, because the combination is **order-dependent** — measured on `${m[(FLAG)a,b]}`, zsh 5.9.2:

| Flag | Native |
| --- | --- |
| `(i)`, `(I)` | `invalid subscript` |
| `(ir)` | valid |
| `(ri)` | **`invalid subscript`** |
| `(ik)` | valid |
| `(in)` | `bad math expression` (the #390 family) |
| `(r)`, `(k)` | valid |

`(ir)` and `(ri)` are the same letters and disagree, so "the combination's return kind" is not quite it either: the order decides which flag governs. Worth pinning that in whichever issue survives, since a rule keyed on set-membership would accept `(ri)`.

## A blocker that applies to both

The verdict depends on the variable's declared **type**, which the front end cannot see. Same subscript text, zsh 5.9.2:

```console
$ typeset -A v=( a1 x abc y );   print ${v[(i)abc,3]}
                                    # runs clean, rc=0

$ local -a v=( a1 abx abc dd );  print ${v[(i)abc,3]}
t_array.zsh:3: invalid subscript    # rc=1
```

| Declared type of `v` | `${v[(i)abc,3]}` |
| --- | --- |
| `local -a` (array) | `invalid subscript` |
| scalar | `invalid subscript` |
| undeclared | `invalid subscript` |
| `typeset -A` (assoc) | **valid** |

So a source-only rejection in `internal/parse` would false-reject valid associative code. `internal/scope` tracks only `KindVariable` / `KindFunction` / `KindAlias`, so assoc-vs-array is not available at the analyzer layer either without extending it first.

Three of four declaration states reject, so a type-aware rule could still cover array, scalar and undeclared — most real code — and leave associative subscripts alone. That is the shape I would suggest for whichever of #387/#389 survives the reconciliation.

Details and the 58-row probe are in https://github.com/z-shell/zsh-lint/issues/389#issuecomment-5818727047. The same wall applies to #390 (bare `(n)`/`(b)`/`(P)`), filed today.

**Comment by ss-o** [comment]
Blocked on #391.

This needs associative-vs-normal array tracking in `internal/scope`, which does not exist yet: the native verdict here depends on the subscripted variable's declared type, and `internal/parse` cannot see it. A source-only rule would false-reject valid associative code.

#391 is the enabler (the AST already carries the `-A` flag; `internal/scope` just does not record it). Once it lands, the rule for this issue belongs in an analyzer rule reading the symbol map, not in the parser front end.
